### ISSUE TITLE
No longer able to send non-string attribute values

### ISSUE THREAD

**Issue by actionshrimp** [perezd/riemann-nodejs-client#16]
Hi,

Big fan of the `protocol-buffers` update overall as it brings our `npm install` time down massively as we no longer have to compile anything so thanks for that :). Unfortunately it looks like attributes with values that are not strings are no longer supported - these worked previously. I don't want to convert them to strings as the place they end up handles strings differently for numeric types.

To replicate, you can add something like `{key: "response_time", value: 123}` to the 'custom attributes' test (around line 72) in `test/basic_tests.js` and run the tests - this causes a couple of extra failures, although it looks like mocha eats the stacktrace as done gets called twice.

The stack trace we're seeing at application level is:

```
TypeError: Argument must be a string
    at Object.exports.string.encodingLength (/home/dave/dev/code/webstore/node_modules/riemann/node_modules/protocol-buffers/encodings.js:60:22)
    at Object.encodingLength (eval at <anonymous> (/home/dave/dev/code/webstore/node_modules/riemann/node_modules/protocol-buffers/node_modules/generate-function/index.js:55:21), <ano$
ymous>:8:22)
    at Object.encodingLength (eval at <anonymous> (/home/dave/dev/code/webstore/node_modules/riemann/node_modules/protocol-buffers/node_modules/generate-function/index.js:55:21), <ano$
ymous>:38:24)
    at encodingLength (eval at <anonymous> (/home/dave/dev/code/webstore/node_modules/riemann/node_modules/protocol-buffers/node_modules/generate-function/index.js:55:21), <anonymous>$
28:24)
    at Object.encode (eval at <anonymous> (/home/dave/dev/code/webstore/node_modules/riemann/node_modules/protocol-buffers/node_modules/generate-function/index.js:55:21), <anonymous>:$
:30)
    at _serialize (/home/dave/dev/code/webstore/node_modules/riemann/riemann/serializer.js:11:30)
    at Object.exports.serializeMessage (/home/dave/dev/code/webstore/node_modules/riemann/riemann/serializer.js:30:10)
    at udpSocket.<anonymous> (/home/dave/dev/code/webstore/node_modules/riemann/riemann/client.js:22:30)
    at Client.send (/home/dave/dev/code/webstore/node_modules/riemann/riemann/client.js:136:11)
    at <our app code>
```

Let me know if there's anything else I can do to help diagnose.

Thanks!
Dave

**Comment by perezd** [comment]
I suggest you hard code your dependency to 0.4.x until this is resolved.
I'm accepting pull requests if you'd like to fix this. Thanks!

**Comment by perezd** [comment]
@robertklep Would you mind investigating/patching this? This breakage is directly related to your PR. Thanks!

**Comment by actionshrimp** [comment]
I had a dig actually and it looks like this might now be correct behaviour whereas the old behaviour was exploiting a bug, (even though riemann supported it fine)?

The protobuf spec specifies that the attribute value is a string: https://github.com/perezd/riemann-nodejs-client/blob/master/riemann/proto/proto.proto#L44 . Maybe that spec can be changed, but I imagine the spec comes from riemann itself and I don't know enough about protobufs to know if they support variable types.

**Comment by perezd** [comment]
Riemann probably will continue to work if you serialize it into a string then. Maybe try that? I imagine riemann is duck typing the types it receives.

**Comment by robertklep** [comment]
I think that maybe the previous protocol buffer module would cast an incoming value to match the spec, but I'd have to look into that to make sure (it'll have to wait until tomorrow though, I'm afraid).

**Comment by perezd** [comment]
Yeah, my guess is that it was coercing it, as well.

**Comment by robertklep** [comment]
FWIW, the official spec declares value to be a string: https://github.com/aphyr/riemann-java-client/blob/master/riemann-java-client/src/main/proto/riemann/proto.proto

**Comment by perezd** [comment]
Yes, however people may have been relying on this bug, so perhaps we should add in code that automatically coerces it back to a string for compat reasons.

**Comment by actionshrimp** [comment]
I just compared the old and new behaviour and it does look like the old lib was coercing to string.

**Comment by robertklep** [comment]
@actionshrimp yeah confirmed with the old module, it is coercing to string.

@perezd unless it's possible to introspect the protocul buffer spec (to find string-type field names), that may become very hacky. It may be better to release with an increased major version to signify backward incompatible changes?

**Comment by perezd** [comment]
Well, technically I already did that. this is 0.5.0, the last one was 0.4.2. 

I imagine the logic is, take numbers (likely the only value that riemann understands) and parse them into strings. Right? Maybe bools as well? It's not going to need to support anything other than primitives.

**Comment by robertklep** [comment]
@perezd increasing the major would make it `1.0.0`.

I'll see if I can put together a workaround, although the issue is that such a workaround would require intimate knowledge of the protocol buffer spec, which would—in the event of future changes of the official spec—make it prone to break.

**Comment by perezd** [comment]
This isn't ready to be 1.0.

There's nothing to know about the proto spec. It says its accepts a string, it must be a string, thats it. So the only thing we need to do is call .toString() on whatever the value is we get. I believe its that simple.

**Comment by robertklep** [comment]
@perezd the spec also uses bools, doubles, floats, etc, so we have to known which fields should be of what type before we can cast it properly. I'll see if I can find a replacement protobuf module that can perform the casting automatically.

**Comment by Priyanky** [comment]
@all I noticed one behaviour with version 0.5.0 , The field metric will not work any more. 
And we need to use metric_d /metric_f . But the previous versions were compatible with field metric.
This is not obvious and figured following the lead by  @robertklep  in https://github.com/aphyr/riemann-java-client/blob/master/riemann-java-client/src/main/proto/riemann/proto.proto
@perezd , Can you update the documentation to specify this

**Comment by robertklep** [comment]
@Priyanky if you feel like it, you can try and see if #17 fixes your problem:

```
$ npm install robertklep/riemann-nodejs-client#8bb5fbc100455f1d85409ab7fb970fc21a46a138
```

(it's just a quick fix, eventually it should be merged and published to the NPM repository)

**Comment by perezd** [comment]
@Priyanky this is actually not related to this bug, we do already support metricF:
https://github.com/perezd/riemann-nodejs-client/blob/master/riemann/client.js#L48

Its backwards compatible.

**Comment by perezd** [comment]
@robertklep, @actionshrimp  looks like we've really made a mess of things, haven't we :).

OK, I'll will merge #17 and push it as 0.5.1 and we'll never speak of the 0.5.0 release again. If you guys are still concerned about node version incompatibility, I'm open to a future PR for a different library, but we need better testing.

**Comment by perezd** [comment]
OK This is now released as 0.5.1.

**Comment by Priyanky** [comment]
@perezd In version 0.5.0 metric value was not working for me for the non string values. So worked around by using metric_d/metric_f . Did not checked 0.5.1 yet.
